### ISSUE TITLE
[Feature request] Force state update to be applied (not batched) after callback finishes

### ISSUE THREAD

**Issue by dgrechka** [facebookexperimental/Recoil#1098]
I have the following issue.

Let's say I have atom A with value 1.

1) In mouse event handler the recoil callback with setter is called to set A value to 2.
2) setTimeout callback fires, which in turn contains the recoil callback call. the `snapshot` in it still contains the A value of 1. The result of first step is not applied.

My question.
Can I force all of the updates of the recoil callback to be applied right after the callback exits. e.g explicitly force not to batch it into a single transaction with subsequent callback calls ?

**Comment by drarmstr** [comment]
It sounds like you are asking about the observability of changes within the callback itself, versus the timing changes are applied to the React component tree.  For this, try out the new `useRecoilTransaction_UNSTABLE()` API #1085 in the nightly build.  Other options that may suffice with the current hook are using the updater form of setter by providing a callback which gets the "current" value of the state as a parameter value.  Another option may be calling another callback from `useRecoilCallback()` from the timeout to get an updated `Snapshot`.

**Comment by dgrechka** [comment]
Thank you for the feedback! much appreciated!

Indeed, I need the changes made by one callback to be visible by another callback. Both callbackes are obtained via seperate`useRecoilCallback` calls.

As for the proposed options.
1. Seems that `useRecoilTransaction_UNSTABLE` can help for my use case. Am I right that after callback obtained via `useRecoilTransaction_UNSTABLE` exits, the changes made by it become visible to the subsequent callbacks calls (obtained either via `useRecoilCallback` or via `useRecoilTransaction_UNSTABLE`)?
2. Calling updater instead of providing new value in setter is not an option for me.
In my recoil callback I read the data through one selector, and write back the data into recoil through another selector. Although these selectors are based on the same set of underlying atoms, they provide the data in different forms (types).
3. Could you elaborate on this option, please?
It seems that I already have it but it does not work as expected.

The toy example of my case is the following.

```{typescript}
const callback1 = useRecoilCallback(({snapshot,set}) => ... => {
   ...
   set(atomA, 2);
   console.log('atomA value updated to 2 in the first callback');
   ...
}) // this one is called in mouse event handler,  it is executed first.

const callback2 = useRecoilCallback(({snapshot,set}) => ... => { 
 console.log(`read back atomA value "${snapshot.getLoadable(atomA).GetValue()}" in the second callback`)
}) // this callback is called setTimeout, it fires after the callback1 call within mouse event handler
```

In the console I see
```
atomA value updated to 2 in the first callback
read back atomA value 1 in the second callback
```

**Comment by dgrechka** [comment]
When I added 
```
useRecoilTransactionObserver_UNSTABLE(({ snapshot, previousSnapshot }) => {
    const getVal = (s: Snapshot) => s.getLoadable(atomA).getValue();
    console.log('recoil transaction. atom was ', getVal(previousSnapshot), '. now it is', getVal(snapshot));
  });
```
to better understand when the writes are applied, I can see that 
`recoil transaction. atom was 1, now it is 2`
appears in the console only after both callbacks are executed.
As I understand, what I need is to see this happening in between them.

Note that it is rare case! There is a race condition here. Most of the time I can see the writes of the first callback in the second one. But sometimes they are not applied and I see stale values.

I can imaging the API like `useRecoilForceCommitTransaction` which returns callback of `() => void` type. This callback internally can call [applyActionsToStore](https://github.com/facebookexperimental/Recoil/blob/42ddba22a792d1f602bb1398b2b4eb238c9017b8/src/core/Recoil_RecoilValueInterface.js#L191) (i'm not sure it is correct way to do it) or by other means applies the accumulated updates, so the next `useRecoilCallback` obtained callback call is guaranteed to get the snapshot with all of the pending updates applied.
Does it make sense?

**Comment by dgrechka** [comment]
After doing a lot of tracing I found that [effect](https://github.com/facebookexperimental/Recoil/blob/42ddba22a792d1f602bb1398b2b4eb238c9017b8/src/core/Recoil_RecoilRoot.react.js#L259) that [promotes nextTree to the currentTree inside store](https://github.com/facebookexperimental/Recoil/blob/42ddba22a792d1f602bb1398b2b4eb238c9017b8/src/core/Recoil_RecoilRoot.react.js#L215) may not be called after callback1 exits.
React does not do render immediately after callback1 exists, which is absolutly fine. Maybe it decides to postpone the render to do some state update batching

Meanwhile callback1 prepared the nextTree to be applied. timeout with callback2 fires before the effect runs, thus recoil changes are still held in the nextTree and the callback2 reads the old snapshot (based on stale currentTree).

So what I need is 
1. either ability to force [endBatch](https://github.com/facebookexperimental/Recoil/blob/42ddba22a792d1f602bb1398b2b4eb238c9017b8/src/core/Recoil_RecoilRoot.react.js#L200) explicitly before callback2 (e.g. with `useRecoilForceCommitTransaction` which returns callback of () => void type).
2. or an ability to switch `useCallback` to capture snapshot not based on currentTree, but on the nextTree instead (e.g. add `pendingSnapshot` to the `RecoilCallbackInterface` ).

**Comment by drarmstr** [comment]
It sounds like the `useRecoilTransaction_UNSTABLE()` API #1085 would be a good solution for you, then, if it supported selectors.  This support is part of the vision for Recoil transactions.

**Comment by salvoravida** [comment]
@dgrechka 
"In my recoil callback I read the data through one selector, and write back the data into recoil through another selector."

I think that this is the point: you should not reading from A and setting B, but instead, you should create a new selector C that reactive to that changes.

It's like reading from db while may are pending transactions and updating another db view from that reading. you should correct the view.

**Comment by salvoravida** [comment]
Meanwhile, @drarmstr great job with useRecoilTransaction!

**Comment by yoavniran** [comment]
Am I right to think that `useRecoilTransaction` would/should  be the only way to write setters with multiple "set" calls? If so, its really missing the option to _get_ from selectors (and not just atoms). Is there a release date we can expect this to become available? thanks!
